corporate/lib/stripe.py

```diff
@@ -478,6 +508,42 @@ def invoice_plan(plan: CustomerPlan, event_time: datetime) -> None:
         plan.save(update_fields=['invoicing_status', 'invoiced_through'])
         licenses_base = ledger_entry.licenses
 
+    if plan.status == CustomerPlan.ENDED:
+        new_plan = get_current_plan_by_realm(plan.customer.realm)
+
+        if new_plan is not None and new_plan.invoicing_status == CustomerPlan.INITIAL_INVOICE_TO_BE_SENT:
```

Comment:
Don't we risk double-charging here?
```
def invoice_plans_as_needed(event_time: datetime=timezone_now()) -> None:
    for plan in CustomerPlan.objects.filter(next_invoice_date__lte=event_time):
        invoice_plan(plan, event_time)
```

Could it happen that this for loop will run over the `ENDED` plan frst, charging the user using this code here, and then it will call `invoice_plan` on the new plan and charge for it again? I'm only starting to develop any kind of understanding of this system, so I might be talking nonsense though :grinning:

Reply:
> Could it happen that this for loop will run over the ENDED plan frst, charging the user using this code here, and then it will call invoice_plan on the new plan and charge for it again?

The `new_plan` won't be present in the loop. The loop only has objects with `next_invoice_date__lte=event_time`. The  `next_invoice_date` of the new_plan is set to 1 month/12 month(depending on whether the plan is on automatic or manual license management) from the day of the switch :)

Reply:
Also, we mark `new_plan.invoiced_through = ledger_entry` when we process the old plan. So even if the new_plan object comes after the old_plan is invoiced, there won't be a duplicate charge since only the ledger entries that are greater than `invoiced_through` is invoiced.

Reply:
Ah I see, that sounds right then. In that case perhaps it's not a priority, but wouldn't it be an improvement to structure the logic in such a way that `invoice_plan()` will be called on the new plan, and that will make organization get billed? Rather than it mutating the new plan when being called on the old one, which is a bit counter-intuitive .

Reply:
Yeah. I was considering to do it that way. The reason I didn't do that way was it could have generated two invoices instead of 1 in this case. Don't have any other strong reasons.